### ISSUE TITLE
Low precision and range with love.math.noise

### ISSUE THREAD

**Issue by megagrump** [love2d/love#1579]
`love.math.noise` only gives acceptable results with coordinates < ~200,000. Using larger coordinates shows visible banding/pixellation.

```lua
local noise = love.image.newImageData(800, 600)

for i = 0, 600, 200 do
	noise:mapPixel(function(x, y)
		local n = love.math.noise(i * 1000 + x * .01)
		return n, n, n, 1
	end, i, 0, 200, 150)

	noise:mapPixel(function(x, y)
		local n = love.math.noise(i * 1000 + x * .01, i * 1000 + y * .01)
		return n, n, n, 1
	end, i, 150, 200, 150)

	noise:mapPixel(function(x, y)
		local n = love.math.noise(i * 1000 + x * .01, i * 1000 + y * .01, i * 1000 + y * .01)
		return n, n, n, 1
	end, i, 300, 200, 150)

	noise:mapPixel(function(x, y)
		local n = love.math.noise(i * 1000 + x * .01, i * 1000 + y * .01, i * 1000 + y * .01, i * 1000 + y * .01)
		return n, n, n, 1
	end, i, 450, 200, 150)
end

local img = love.graphics.newImage(noise)

function love.draw()
	love.graphics.draw(img)
end

```

![image](https://user-images.githubusercontent.com/31128870/82528737-bcba7980-9b39-11ea-8c38-e2afa0892446.png)

Also, the range of output values for 1d, 3d and 4d noise is much smaller than it is for 2d noise, resulting in low contrast output.

**Comment by megagrump** [comment]
I wanna fix the aliasing issue at last - the obvious fix is to use `double`s all the way down. I don't think that would be considered a breaking change, would it?

The patent on 3D and 4D noise [expires in a few weeks](https://patents.google.com/patent/US6867776) btw. So we could have SimplexNoise in higher dimensions too in 12.0, if there's no other objection.

**Comment by zorggn** [comment]
Regarding the 3D and 4D ones, can patents be re-filed/renewed? Because if not, then i don't think it'll be an issue come january next year, but if it's a possibility, maybe using OpenSimplex might be a better idea? Not sure if that would have licensing or other issues or not.

**Comment by slime73** [comment]
> the obvious fix is to use doubles all the way down. I don't think that would be considered a breaking change, would it?

I'd consider it breaking because there's a high chance that the output values will change slightly even when the inputs aren't extremely huge. I think it's a good change for 12.0 though.

> we could have SimplexNoise in higher dimensions too in 12.0

There are a few different noise algorithms each with slightly different pros and cons and typical use cases - I wonder if we should redesign the APIs to allow for more choice. Or maybe keeping it simple (and sticking with just one) is best for now, I'm not sure.

**Comment by EngineerSmith** [comment]
I personally would go for simple, aim for an algorithm for games(speed over some precision loss) that can solve the contrast issue above. If someone wants to go into the deep end and want a specific one for their use case they can add it via ffi or implemented it in lua for their project

**Comment by megagrump** [comment]
I'm only gonna fix the aliasing for now.

There's currently 2 types of noise in LÖVE: Simplex for 1D and 2D, and Perlin for 3D and 4D noise. The low contrast is inherent to Perlin; I looked at histograms for every dimension and it turns out that 1D does not actually give a low range of output values. My first assessment was wrong.

If I were to make an extended API, it would look like this (pseudo-Lua):
```lua
local noiseGen = love.math.newNoiseGenerator('simplex' | 'perlin', seed = <some default value>)
local n = noiseGen:noise(x, ...)
```

The seed value is used to initialize an RNG, which generates static offsets. These offsets are simply added to the values passed to `:noise` - an answer to the (mostly hypothetical) question "how to seed the noise generator?"

One can choose between Perlin and Simplex independently from dimensionality, it allows for simply seeding, and it is extensible. Multi-octave noise would be nice to have too. But I'm not sure if that's still considered "basic" functionality, and whether  the additional parameters should be properties of the generator object, or arguments to the generator function.

Again, I'm not doing this right now, it's just a proposal for your consideration for later implementation.

**Comment by EngineerSmith** [comment]
Personally, I like to seed a noise algorithm with an additional dimension than using an offset. I think your approach is fine as long as it is documented what it truly does. If someone wanted to make a 2d world with the noise and thought seed made it unique then they would run into issues. E.g. seed 0 is the same as seed 200 but if they move -200 in X and Y the world would be 100% the sam as seed 0.

Otherwise, you could leave the idea of a seed out, and put a code snippet on the wiki to explain different ways to get repeatable unique values. I do like the idea of choosing a generator using an enum as it is extendable in the future and follows other function conventions

**Comment by megagrump** [comment]
> E.g. seed 0 is the same as seed 200 but if they move -200 in X and Y the world would be 100% the sam as seed 0.

That's why the seed is not used directly as an offset; it is fed to an RNG first, which generates the offsets.

**Comment by megagrump** [comment]
Shuffling the permutation table is another way to add seeding, albeit a slow one. I don't know if that has any other drawbacks.

I got time by the end of this month and could start working on an implementation of my proposal, as a way for me to get more acquainted with the codebase. Maybe sans seeding for now, as it's really not that important. Just say the word, LÖVE göds.

**Comment by MikuAuahDark** [comment]
I think the original issue has been fixed with #1744 and the rest of this conversation talks about the redesign of the LOVE noise APIs and/or using alternative noise functions.

Since the main issue is fixed, I'll close this.
